### PR TITLE
Fix for Conditional Breakpoint not working in File class.

### PR DESCRIPTION
**Pull request by SougandhS**
## What it does
Handles all the exception caused while evaluating conditional breakpoint java.io.File class 
Fix for -> https://github.com/eclipse-jdt/eclipse.jdt.debug/issues/273

## How to test

Drop a valid conditional breakpoint inside java.io.File class and run in debug mode

## Author checklist

- [x] I have thoroughly tested my changes
- [x] The change is following the [coding conventions](https://wiki.eclipse.org/Platform/How_to_Contribute#Coding_Conventions)
- [x] I have signed the [Eclipse Contributor Agreement (ECA)](https://www.eclipse.org/legal/ECA.php)

### CODE REVIEW

**Comment by iloveeclipse** [comment]
Would be very nice to have a regression test added.

File: org.eclipse.jdt.debug.tests/tests/org/eclipse/jdt/debug/tests/AbstractDebugTest.java [hunk 1]
```diff
@@ -184,7 +184,10 @@ public abstract class AbstractDebugTest extends TestCase implements  IEvaluation
 	public static final String BOUND_JRE_PROJECT_NAME = "BoundJRE";
 	public static final String CLONE_SUFFIX = "Clone";
 
-	final String[] LAUNCH_CONFIG_NAMES_1_4 = {"LargeSourceFile", "LotsOfFields", "Breakpoints", "InstanceVariablesTests", "LocalVariablesTests", "LocalVariableTests2", "StaticVariablesTests",
+	final String[] LAUNCH_CONFIG_NAMES_1_4 = { "java.io.File", "FileConditionSnippet2", "LargeSourceFile", "LotsOfFields",
```
Comment (iloveeclipse) [inline org.eclipse.jdt.debug.tests/tests/org/eclipse/jdt/debug/tests/AbstractDebugTest.java:187]: `"java.io.File",` looks wrong. Also please don't reformat every line, only add new test at the end

File: org.eclipse.jdt.debug.tests/tests/org/eclipse/jdt/debug/tests/AbstractDebugTest.java [hunk 2]
```diff
@@ -1536,13 +1539,14 @@ protected IJavaThread launchToLineBreakpoint(String mainTypeName, ILineBreakpoin
 	 */
 	protected IJavaThread launchToLineBreakpoint(ILaunchConfiguration config, ILineBreakpoint bp, boolean register) throws Exception {
 		DebugEventWaiter waiter= new DebugElementKindEventDetailWaiter(DebugEvent.SUSPEND, IJavaThread.class, DebugEvent.BREAKPOINT);
-		waiter.setTimeout(DEFAULT_TIMEOUT);
+		waiter.setTimeout(30000);
```
Comment (iloveeclipse) [inline org.eclipse.jdt.debug.tests/tests/org/eclipse/jdt/debug/tests/AbstractDebugTest.java:1542]: Why is this changed from `DEFAULT_TIMEOUT` to `30000`? Please restore. This will affect every other test too.

File: org.eclipse.jdt.debug.tests/tests/org/eclipse/jdt/debug/tests/AbstractDebugTest.java [hunk 3]
```diff
@@ -1536,13 +1539,14 @@ protected IJavaThread launchToLineBreakpoint(String mainTypeName, ILineBreakpoin
...
 		waiter.setEnableUIEventLoopProcessing(enableUIEventLoopProcessingInWaiter());
 
 		Object suspendee= launchAndWait(config, waiter, register);
 		assertTrue("suspendee was not an IJavaThread", suspendee instanceof IJavaThread); //$NON-NLS-1$
 		IJavaThread thread = (IJavaThread) suspendee;
 		IBreakpoint hit = getBreakpoint(thread);
+
```
Comment (iloveeclipse) [inline org.eclipse.jdt.debug.tests/tests/org/eclipse/jdt/debug/tests/AbstractDebugTest.java:1549]: Please avoid unnecessary white space changes

File: org.eclipse.jdt.debug.tests/testprograms/FileConditionSnippet2.java [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+	public static void main(String[] ecs) {
+		int i = 0;
+		File parent = new File("parent");
+		File file = new File(parent,"test");	
+		System.out.println("COMPLETED");
+	}
+
+}
+
+/////////////////////////////////////////
```
Comment (iloveeclipse) [inline org.eclipse.jdt.debug.tests/testprograms/FileConditionSnippet2.java:13]: Please remove this comment line, it is not needed

File: org.eclipse.jdt.debug.tests/tests/org/eclipse/jdt/debug/tests/breakpoints/ConditionalBreakpointsWithFileClass.java
Comment (iloveeclipse) [inline org.eclipse.jdt.debug.tests/tests/org/eclipse/jdt/debug/tests/breakpoints/ConditionalBreakpointsWithFileClass.java:1]: ConditionalBreakpointsWithFileClass should be added to the testsuite

**Review by iloveeclipse — CHANGES_REQUESTED** [review]
Thanks for the new test, there are few smaller issues to fix, see comments above

**Comment by iloveeclipse** [comment]
Also please **rebase** (do **not** merge with!) on latest master branch state.

File: org.eclipse.jdt.debug.tests/testprograms/FileConditionSnippet2.java (hunk 4 above)
Comment (SougandhS) [reply org.eclipse.jdt.debug.tests/testprograms/FileConditionSnippet2.java:13]: Changes done.

File: org.eclipse.jdt.debug.tests/tests/org/eclipse/jdt/debug/tests/AbstractDebugTest.java (hunk 2 above)
Comment (SougandhS) [reply org.eclipse.jdt.debug.tests/tests/org/eclipse/jdt/debug/tests/AbstractDebugTest.java:1542]: Restored to previous.

File: org.eclipse.jdt.debug.tests/tests/org/eclipse/jdt/debug/tests/AbstractDebugTest.java (hunk 3 above)
Comment (SougandhS) [reply org.eclipse.jdt.debug.tests/tests/org/eclipse/jdt/debug/tests/AbstractDebugTest.java:1549]: Removed.

File: org.eclipse.jdt.debug.tests/tests/org/eclipse/jdt/debug/tests/breakpoints/ConditionalBreakpointsWithFileClass.java
Comment (SougandhS) [reply org.eclipse.jdt.debug.tests/tests/org/eclipse/jdt/debug/tests/breakpoints/ConditionalBreakpointsWithFileClass.java:1]: Could you pls advice me on it ?

Comment (SougandhS) [reply org.eclipse.jdt.debug.tests/tests/org/eclipse/jdt/debug/tests/breakpoints/ConditionalBreakpointsWithFileClass.java:1]: Added now.

Comment (SougandhS) [reply org.eclipse.jdt.debug.tests/tests/org/eclipse/jdt/debug/tests/breakpoints/ConditionalBreakpointsWithFileClass.java:1]: The unit tests are passing in local but failing in jenkins automated tests.

Comment (SougandhS) [reply org.eclipse.jdt.debug.tests/tests/org/eclipse/jdt/debug/tests/breakpoints/ConditionalBreakpointsWithFileClass.java:1]: @iloveeclipse on jenkins 
-> 
image

on eclipse -> 
image
Do we have to include any additional changes to work in Jenkins for File class?

File: org.eclipse.jdt.debug.tests/tests/org/eclipse/jdt/debug/tests/breakpoints/ConditionalBreakpointsWithFileClass.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+			// System.out.println("--> " + mainThread.getTopStackFrame().getLineNumber());
+			Thread.sleep(100);
+			mainThread.resume();
+			Thread.sleep(100);
+			// System.out.println("--> " + mainThread.getTopStackFrame().getLineNumber());
+			int hitLine = 0;
+			if (mainThread.isSuspended()) {
+				hitLine = mainThread.getStackFrames()[0].getLineNumber();
+			}
+			assertTrue("Should've hit on line 374 but hit on " + hitLine, (hitLine == 374));
```
Comment (iloveeclipse) [inline org.eclipse.jdt.debug.tests/tests/org/eclipse/jdt/debug/tests/breakpoints/ConditionalBreakpointsWithFileClass.java:30]: Two things: why do you have `if (mainThread.isSuspended())` as condition? It should be an assert.
So you will see whether the thread is suspended or not.
Second: please instead of `assertTrue(x == y)` please use **assertEquals**.

Both will not explain the fail but would help to understand the root cause.

Comment (SougandhS) [reply org.eclipse.jdt.debug.tests/tests/org/eclipse/jdt/debug/tests/breakpoints/ConditionalBreakpointsWithFileClass.java:30]: Will apply these changes and check again.

Comment (SougandhS) [reply org.eclipse.jdt.debug.tests/tests/org/eclipse/jdt/debug/tests/breakpoints/ConditionalBreakpointsWithFileClass.java]: @iloveeclipse Tried out several approaches and still fails in Jenkins ->
image

Is it because we are invoking java inbuilt class during test ?

Comment (SougandhS) [reply org.eclipse.jdt.debug.tests/tests/org/eclipse/jdt/debug/tests/breakpoints/ConditionalBreakpointsWithFileClass.java]: @iloveeclipse Done all the changes and tested, kindly review the changes

**Comment by SougandhS** [comment]
Hi @iloveeclipse, done all the changes.

**Comment by iloveeclipse** [comment]
> Hi @iloveeclipse, done all the changes.

Sorry for me not answering timely, I'm currently working 2 days in a week only and have almost no free time for PR reviews.

**Comment by SougandhS** [comment]
> > > Hi @iloveeclipse, done all the changes.
> > 
> > 
> > Sorry for me not answering timely, I'm currently working 2 days in a week only and have almost no free time for PR reviews.
> 
> Okay @iloveeclipse ,no problem.

Hi @jukzi ,could you pls review this ?

File: org.eclipse.jdt.debug/eval/org/eclipse/jdt/internal/debug/eval/ast/engine/ASTEvaluationEngine.java [hunk 6]
```diff
@@ -92,30 +92,24 @@ public class ASTEvaluationEngine implements IAstEvaluationEngine {
...
 		@Override
 		public DebugEvent[] filterDebugEvents(DebugEvent[] events) {
 			if (events.length == 1) {
 				DebugEvent event = events[0];
-				if (event.getSource() instanceof IJavaVariable
-						&& event.getKind() == DebugEvent.CHANGE) {
-					if (((IJavaVariable) event.getSource()).getDebugTarget()
-							.equals(getDebugTarget())) {
+				if (event.getSource() instanceof IJavaVariable && event.getKind() == DebugEvent.CHANGE) {
+					if (((IJavaVariable) event.getSource()).getDebugTarget().equals(getDebugTarget())) {
```
Comment (jukzi) [inline org.eclipse.jdt.debug/eval/org/eclipse/jdt/internal/debug/eval/ast/engine/ASTEvaluationEngine.java:112]: avoid typecast with instanceof pattern

**Review by jukzi — COMMENTED** [review]
With all that code reformat it is hard to identify in the review what the actual change is. but overall it looks sane.

File: org.eclipse.jdt.debug/model/org/eclipse/jdt/internal/debug/core/breakpoints/ConditionalBreakpointHandler.java [hunk 7]
```diff
@@ -217,12 +220,15 @@ public int breakpointHit(IJavaThread thread, IJavaBreakpoint breakpoint) {
 						.getTopStackFrame();
 				IJavaProject project = lineBreakpoint.getJavaProject(frame);
 				if (project == null) {
-					fireConditionHasErrors(
-							lineBreakpoint,
-							new Message[] { new Message(
-									JDIDebugBreakpointMessages.JavaLineBreakpoint_Unable_to_compile_conditional_breakpoint___missing_Java_project_context__1,
-									-1) });
-					return SUSPEND;
+					IMarker marker = breakpoint.getMarker(); // my changes
```
Comment (jukzi) [inline org.eclipse.jdt.debug/model/org/eclipse/jdt/internal/debug/core/breakpoints/ConditionalBreakpointHandler.java:223]:
```suggestion
					IMarker marker = breakpoint.getMarker();
```

File: org.eclipse.jdt.debug/model/org/eclipse/jdt/internal/debug/core/breakpoints/ConditionalBreakpointHandler.java [hunk 8]
```diff
@@ -217,12 +220,15 @@ public int breakpointHit(IJavaThread thread, IJavaBreakpoint breakpoint) {
...
+					if (marker != null) {
+						IResource res = marker.getResource();
+						project = JavaCore.create(res.getProject());
```
Comment (jukzi) [inline org.eclipse.jdt.debug/model/org/eclipse/jdt/internal/debug/core/breakpoints/ConditionalBreakpointHandler.java:226]: the javadoc of org.eclipse.core.resources.IResource.getProject() says it may retun null. Better catch that case?

**Comment by jukzi** [comment]
Is it possible with adequate effort to split the automated code reformating in a separate commit (Ok to have it in the same PR though)  to make the review easier? For that you could just reformat the classes in question and rebase this PR on top of it.

File: org.eclipse.jdt.debug/eval/org/eclipse/jdt/internal/debug/eval/ast/engine/ASTEvaluationEngine.java (hunk 6 above)
Comment (SougandhS) [reply org.eclipse.jdt.debug/eval/org/eclipse/jdt/internal/debug/eval/ast/engine/ASTEvaluationEngine.java:112]: Will change those now.

File: org.eclipse.jdt.debug/model/org/eclipse/jdt/internal/debug/core/breakpoints/ConditionalBreakpointHandler.java (hunk 8 above)
Comment (SougandhS) [reply org.eclipse.jdt.debug/model/org/eclipse/jdt/internal/debug/core/breakpoints/ConditionalBreakpointHandler.java:226]: Sure

**Comment by SougandhS** [comment]
> Is it possible with adequate effort to split the automated code reformating in a separate commit (Ok to have it in the same PR though) to make the review easier? For that you could just reformat the classes in question and rebase this PR on top of it.

Sure will do that.

**Comment by jukzi** [comment]
@RoiSoleil can you test if this PR solves the issue you raised?

**Comment by SougandhS** [comment]
> With all that code reformat it is hard to identify in the review what the actual change is. but overall it looks sane.


> Is it possible with adequate effort to split the automated code reformating in a separate commit (Ok to have it in the same PR though) to make the review easier? For that you could just reformat the classes in question and rebase this PR on top of it.

Actually the changes are only in ASTEvaluationEngine.createExpressionFromAST(String,EvaluationSourceGenerator,CompilationUnit) line: 601 
(Added additional problem id filter for Ambiguity)

ConditionalBreakpointHandler.breakpointHit(IJavaThread,IJavaBreakpoint) line: 222
(Getting project details if provided with BinaryType)

Rest of changes are only for test cases.

**Comment by RoiSoleil** [comment]
> @RoiSoleil can you test if this PR solves the issue you raised?

It's working with the PR. Ok for me.

**Comment by jukzi** [comment]
looks ok, was tested => lets merge

**Comment by SougandhS** [comment]
Thank you @jukzi
